packages/cli/src/ui/server.ts

```diff
@@ -0,0 +1,187 @@
+    supportedProtocols: [ProtocolId.HistoryNetwork],
+  })
+  portal.discv5.enableLogs()
+  portal.enableLog('*')
+
+  const history = portal.protocols.get(ProtocolId.HistoryNetwork) as HistoryProtocol
+  const router = t.router
+
+  //  WSS Client Methods
+
+  const decodeENR = publicProcedure
+    .input(z_Enr)
+    .output(
+      z.object({
+        nodeId: z.string(),
+        multiaddr: z.string(),
+        c: z.string(),
+      }),
+    )
+    .mutation(({ input }) => {
+      const enr = ENR.decodeTxt(input)
+      const tag = enr.kvs.get('c')
+      const c = tag ? tag.toString() : ''
+      const nodeId = enr.nodeId
+      const multiaddr = enr.getLocationMultiaddr('udp')?.toString() ?? ''
+      return {
+        nodeId,
+        multiaddr,
+        c,
+      }
+    })
+
+  const { onTalkReq, onTalkResp, onContentAdded, onNodeAdded, onSendTalkReq, onSendTalkResp } =
+    await subscriptions(portal, history, publicProcedure)
+
+  const {
+    browser_nodeInfo,
+    browser_localRoutingTable,
+    ping,
+    pingBootNodes,
+    browser_historyStore,
+    browser_historyLocalContent,
+  } = websocketProcedures(portal, publicProcedure)
+
+  /**
+   * HTTP Client Methods
+   */
+
+  const {
+    portal_historyRoutingTableInfo,
+    local_routingTable,
+    portal_historyPing,
+    discv5_nodeInfo,
+    pingBootNodeHTTP,
+    portal_historyStore,
+    getPubIp,
+    portal_historyGetEnr,
+  } = httpProcedures(publicProcedure, pubIp)
+
+  // Create tRpc Router
+
+  const appRouter = router({
+    getPubIp,
+    decodeENR,
+    onTalkReq,
+    onTalkResp,
+    onSendTalkReq,
+    onSendTalkResp,
+    onContentAdded,
+    onNodeAdded,
+    browser_nodeInfo,
+    local_routingTable,
+    browser_localRoutingTable,
+    ping,
+    pingBootNodes,
+    discv5_nodeInfo,
+    portal_historyRoutingTableInfo,
+    portal_historyGetEnr,
+    portal_historyPing,
+    portal_historyStore,
+    browser_historyStore,
+    browser_historyLocalContent,
+    pingBootNodeHTTP,
+  })
+
+  // export only the type definition of the API
+  // None of the actual implementation is exposed to the client
+
+  const wss = new ws.Server({
+    port: 3001,
+  })
+  const handler = applyWSSHandler({
+    wss,
+    router: appRouter,
+    createContext() {
+      console.log('context 4')
+      return {}
+    },
+  })
+  wss.on('connection', (ws) => {
+    console.log(`➕➕ Connection (${wss.clients.size})`)
+    ws.once('close', () => {
+      console.log(`➖➖ Connection (${wss.clients.size})`)
+    })
+  })
+  wss.on('error', (err: any) => {
+    console.error(`❌ WebSocket Server error: ${err.message}`)
+  })
+  console.log('✅ WebSocket Server listening on ws://localhost:3001')
+  process.on('SIGTERM', () => {
+    console.log('SIGTERM')
+    handler.broadcastReconnectNotification()
+    wss.close()
+  })
+  portal.discv5.enableLogs()
+
+  portal.enableLog('*ultralight*, *LightClient*, -OFFER, -ACCEPT, *ultralight:RPC*')
+
+  await portal.start()
```

Comment:
Just to make sure I understand what this is doing, this is just starting an Ultralight client and relaying the messages over WSS to the UI, right?

Reply:
Yes -- this part starts an Ultralight client inside of the server, where we can hook directly into the typescript, and utilize event listeners.

This is unique to the "HTTP" client mode, which communicates with a CLI node running independently

Reply:
Ultimately i should probably decouple this, so that running the internal Ultralight instance is optional, or an entirely different script.